From: Jaldhar H. Vyas <jaldhar@debian.org>
Date: Wed, 18 Jul 2012 17:35:35 +0300
Subject: backport of 2.1.8 fixes for mailbox names
Bug: #680035
Description: Fixes instances where UTF-8 was misused instead of mUTF-7
 This patch is an amalgamation of the following HG changesets
 # HG changeset patch
 # User Timo Sirainen <tss@iki.fi>
 # Date 1341230073 -10800
 # Node ID a698fccd37c3cf82fef4e5e9f9a3f5c6a983e437
 # Parent  569588ff7ef07f70a1f7c18c9063db391cb18873
 imap: Mailbox names in STATUS replies were sent as UTF-8 instead of mUTF-7.

 # HG changeset patch
 # User Timo Sirainen <tss@iki.fi>
 # Date 1341266770 -10800
 # Node ID 880af3c78df9adb61fe012792ce0e48bcd08752b
 # Parent  a698fccd37c3cf82fef4e5e9f9a3f5c6a983e437
 imap-acl: MYRIGHTS command used UTF-8 instead of mUTF-7 encoding for mailbox name.

 # HG changeset patch
 # User Timo Sirainen <tss@iki.fi>
 # Date 1341266784 -10800
 # Node ID 90270c054e1f64548e3ce5d73db85739ba9ab5f8
 # Parent  880af3c78df9adb61fe012792ce0e48bcd08752b
 imap-quota: GETQUOTAROOT command used UTF-8 instead of mUTF-7 encoding for mailbox name.
Origin: http://hg.dovecot.org/dovecot-2.1
 
---

diff -r 569588ff7ef0 -r a698fccd37c3 src/imap/cmd-list.c
--- a/src/imap/cmd-list.c	Mon Jul 02 10:12:59 2012 +0300
+++ b/src/imap/cmd-list.c	Mon Jul 02 14:54:33 2012 +0300
@@ -364,8 +364,9 @@
 	client_send_line(ctx->cmd->client, str_c(str));
 }
 
-static void list_send_status(struct cmd_list_context *ctx, const char *name,
-			     enum mailbox_info_flags flags)
+static void
+list_send_status(struct cmd_list_context *ctx, const char *name,
+		 const char *mutf7_name, enum mailbox_info_flags flags)
 {
 	struct imap_status_result result;
 	struct mail_namespace *ns;
@@ -391,7 +392,8 @@
 		return;
 	}
 
-	imap_status_send(ctx->cmd->client, name, &ctx->status_items, &result);
+	imap_status_send(ctx->cmd->client, mutf7_name,
+			 &ctx->status_items, &result);
 }
 
 static bool list_has_empty_prefix_ns(struct mail_user *user)
@@ -481,7 +483,7 @@
 
 		ret = client_send_line(ctx->cmd->client, str_c(str));
 		if (ctx->used_status) T_BEGIN {
-			list_send_status(ctx, name, flags);
+			list_send_status(ctx, name, str_c(mutf7_name), flags);
 		} T_END;
 		if (ret == 0) {
 			/* buffer is full, continue later */
diff -r 569588ff7ef0 -r a698fccd37c3 src/imap/cmd-status.c
--- a/src/imap/cmd-status.c	Mon Jul 02 10:12:59 2012 +0300
+++ b/src/imap/cmd-status.c	Mon Jul 02 14:54:33 2012 +0300
@@ -13,7 +13,7 @@
 	struct imap_status_items items;
 	struct imap_status_result result;
 	struct mail_namespace *ns;
-	const char *mailbox, *error;
+	const char *mailbox, *orig_mailbox, *error;
 	bool selected_mailbox;
 
 	/* <mailbox> <status items> */
@@ -30,6 +30,7 @@
 	if (imap_status_parse_items(cmd, list_args, &items) < 0)
 		return TRUE;
 
+	orig_mailbox = mailbox;
 	ns = client_find_namespace(cmd, &mailbox);
 	if (ns == NULL)
 		return TRUE;
@@ -42,7 +43,7 @@
 		return TRUE;
 	}
 
-	imap_status_send(client, mailbox, &items, &result);
+	imap_status_send(client, orig_mailbox, &items, &result);
 	if (!selected_mailbox)
 		client_send_tagline(cmd, "OK Status completed.");
 	else {
diff -r 569588ff7ef0 -r a698fccd37c3 src/imap/imap-status.c
--- a/src/imap/imap-status.c	Mon Jul 02 10:12:59 2012 +0300
+++ b/src/imap/imap-status.c	Mon Jul 02 14:54:33 2012 +0300
@@ -96,7 +96,7 @@
 	return ret;
 }
 
-void imap_status_send(struct client *client, const char *mailbox,
+void imap_status_send(struct client *client, const char *mailbox_mutf7,
 		      const struct imap_status_items *items,
 		      const struct imap_status_result *result)
 {
@@ -106,7 +106,7 @@
 
 	str = t_str_new(128);
 	str_append(str, "* STATUS ");
-        imap_quote_append_string(str, mailbox, FALSE);
+        imap_quote_append_string(str, mailbox_mutf7, FALSE);
 	str_append(str, " (");
 
 	prefix_len = str_len(str);
diff -r 569588ff7ef0 -r a698fccd37c3 src/imap/imap-status.h
--- a/src/imap/imap-status.h	Mon Jul 02 10:12:59 2012 +0300
+++ b/src/imap/imap-status.h	Mon Jul 02 14:54:33 2012 +0300
@@ -18,7 +18,7 @@
 		    struct mail_namespace *ns,
 		    const char *mailbox, const struct imap_status_items *items,
 		    struct imap_status_result *result_r, const char **error_r);
-void imap_status_send(struct client *client, const char *mailbox,
+void imap_status_send(struct client *client, const char *mailbox_mutf7,
 		      const struct imap_status_items *items,
 		      const struct imap_status_result *result);
 
diff -r a698fccd37c3 -r 880af3c78df9 src/plugins/imap-acl/imap-acl-plugin.c
--- a/src/plugins/imap-acl/imap-acl-plugin.c	Mon Jul 02 14:54:33 2012 +0300
+++ b/src/plugins/imap-acl/imap-acl-plugin.c	Tue Jul 03 01:06:10 2012 +0300
@@ -309,12 +309,13 @@
 {
 	struct mail_namespace *ns;
 	struct mailbox *box;
-	const char *mailbox;
+	const char *mailbox, *orig_mailbox;
 	const char *const *rights;
 	string_t *str;
 
 	if (!client_read_string_args(cmd, 1, &mailbox))
 		return FALSE;
+	orig_mailbox = mailbox;
 
 	if (ACL_USER_CONTEXT(cmd->client->user) == NULL) {
 		client_send_command_error(cmd, "ACLs disabled.");
@@ -346,7 +347,7 @@
 
 	str = t_str_new(128);
 	str_append(str, "* MYRIGHTS ");
-	imap_quote_append_string(str, mailbox, FALSE);
+	imap_quote_append_string(str, orig_mailbox, FALSE);
 	str_append_c(str,' ');
 	imap_acl_write_rights_list(str, rights);
 
diff -r 880af3c78df9 -r 90270c054e1f src/plugins/imap-quota/imap-quota-plugin.c
--- a/src/plugins/imap-quota/imap-quota-plugin.c	Tue Jul 03 01:06:10 2012 +0300
+++ b/src/plugins/imap-quota/imap-quota-plugin.c	Tue Jul 03 01:06:24 2012 +0300
@@ -74,12 +74,13 @@
 	struct mailbox *box;
 	struct quota_root_iter *iter;
         struct quota_root *root;
-	const char *mailbox, *name;
+	const char *mailbox, *orig_mailbox, *name;
 	string_t *quotaroot_reply, *quota_reply;
 
 	/* <mailbox> */
 	if (!client_read_string_args(cmd, 1, &mailbox))
 		return FALSE;
+	orig_mailbox = mailbox;
 
 	ns = client_find_namespace(cmd, &mailbox);
 	if (ns == NULL)
@@ -101,7 +102,7 @@
 	quotaroot_reply = t_str_new(128);
 	quota_reply = t_str_new(256);
 	str_append(quotaroot_reply, "* QUOTAROOT ");
-	imap_quote_append_string(quotaroot_reply, mailbox, FALSE);
+	imap_quote_append_string(quotaroot_reply, orig_mailbox, FALSE);
 
 	iter = quota_root_iter_init(box);
 	while ((root = quota_root_iter_next(iter)) != NULL) {

